File: adversarial-harness/python/policy_engine/timing_guard.py
"""Timing side-channel probe detection — mirrors timing-guard.ts."""

from __future__ import annotations

import hashlib
import os
import re
import time
from typing import Optional

from .arg_walker import walk_string_leaves
from .normalizer import deobfuscate_recursive
from .types import CallContext, PolicyDecision

TIMING_PROBE_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"\b(?:sleep|benchmark|pg_sleep|pg_sleep_for|pg_sleep_until|waitfor\s+delay|"
        r"dbms_lock\.sleep|dbms_pipe\.receive_message|dbms_session\.sleep)\s*\(",
        re.I,
    ),
    re.compile(r"\bif\s*\(\s*(?:ascii|ord|substring|substr|mid|left|right)\s*\(", re.I),
    re.compile(
        r"\b(?:case\s+when|elt\s*\(|decode\s*\()\s+.*\b(?:sleep|benchmark|waitfor|pg_sleep)",
        re.I,
    ),
    re.compile(r"\b(?:timing|time[- ]?based)\s+(?:attack|oracle|injection|blind)", re.I),
    re.compile(r"\b(?:measure|detect|compare)\s+(?:response\s+)?time\s+(?:of|for|between)", re.I),
    re.compile(r"\b(?:valid|invalid)\s+username\b.*\b(?:timing|delay|sleep|benchmark)", re.I),
    re.compile(r"\busername\s+exists\b.*\b(?:time|delay|benchmark|sleep)", re.I),
    re.compile(r"\b(?:SLEEP|BENCHMARK)\s*\(\s*\d+", re.I),
    re.compile(r"\bWAITFOR\s+DELAY\s+'", re.I),
    re.compile(r"\b(?:select|union)\b.+\bwhere\b.+\b(?:sleep|benchmark|waitfor)", re.I),
    re.compile(r"\b'\s*or\s*'1'\s*=\s*'1\b.+\b(?:sleep|benchmark|waitfor|delay)", re.I),
    re.compile(r"\b(?:and|or)\s+\d+\s*=\s*\d+\s*--", re.I),
    re.compile(r"\bldap[_-]?search\b.+\b(?:delay|sleep|time)", re.I),
    re.compile(r"\$where\b.+\b(?:sleep|this\.constructor)", re.I),
    re.compile(r"\b(?:load_file|into\s+outfile)\b.+\b(?:sleep|benchmark)", re.I),
    re.compile(r"\bhex\s*\(\s*(?:substring|mid)\s*\(", re.I),
    re.compile(r"\b(?:response|elapsed)\s+time\s*(?:>|<|>=|<=)\s*\d+", re.I),
    re.compile(r"\buser[_-]?enumeration\b.*\b(?:timing|oracle)", re.I),
    re.compile(r"\b(?:binary|blind)\s+search\b.*\b(?:char|password|secret)", re.I),
]

USERNAME_ORACLE_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\bwhere\b\s+(?:user|username|email|login)\s*=\s*['\"][^'\"]{1,64}['\"]", re.I),
    re.compile(r"\b(?:admin|root|administrator)\b.*\b(?:and|or)\b.*\b(?:sleep|benchmark|waitfor)", re.I),
    re.compile(r"\b(?:'|%27)\s*(?:or|and)\s*(?:'|%27)?\d+['\"]?\s*=\s*['\"]?\d+", re.I),
    re.compile(r"\b(?:exists|in\s*\(\s*select)\b.+\b(?:users|accounts|credentials)", re.I),
]

PROBE_WINDOW_S = 60.0
ENUM_WINDOW_S = 120.0
MAX_TIMING_PROBES = int(os.environ.get("MASTYFF_AI_MAX_TIMING_PROBES_PER_MIN", "8"))
MAX_ENUM_PROBES = int(os.environ.get("MASTYFF_AI_MAX_ENUM_PROBES_PER_SESSION", "20"))

_probe_counters: dict[str, tuple[int, float]] = {}
_enum_counters: dict[str, tuple[int, float]] = {}


def is_timing_guard_enabled() -> bool:
    return os.environ.get("MASTYFF_AI_TIMING_GUARD") != "false"


def reset_timing_probe_counters() -> None:
    _probe_counters.clear()
    _enum_counters.clear()


def _stable_fingerprint(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def scan_timing_probe_patterns(blob: str) -> tuple[bool, list[str]]:
    rule_ids: list[str] = []
    if not blob.strip():
        return False, rule_ids
    for i, pat in enumerate(TIMING_PROBE_PATTERNS):
        if pat.search(blob):
            rule_ids.append(f"timing-probe-{i}")
    for i, pat in enumerate(USERNAME_ORACLE_PATTERNS):
        if pat.search(blob):
            rule_ids.append(f"username-oracle-{i}")
    return len(rule_ids) > 0, rule_ids


def enumeration_fingerprint(blob: str) -> str:
    normalized = blob.lower()
    normalized = re.sub(r"['\"][^'\"]{1,80}['\"]", "<q>", normalized)
    normalized = re.sub(r"\b(?:admin|root|administrator|user\d*|test\d*|guest)\b", "<u>", normalized)
    normalized = re.sub(r"\d+", "N", normalized)
    return _stable_fingerprint(normalized)[:16]


def _increment_counter(
    store: dict[str, tuple[int, float]],
    key: str,
    window_s: float,
) -> int:
    now = time.time()
    count, reset_at = store.get(key, (0, now + window_s))
    if now > reset_at:
        count, reset_at = 1, now + window_s
    else:
        count += 1
    store[key] = (count, reset_at)
    return count


def _build_timing_blob(ctx: CallContext) -> str:
    return "\n".join(
        deobfuscate_recursive(leaf.value) for leaf in walk_string_leaves(ctx.arguments or {})
    )


def _is_enumeration_probe_candidate(blob: str, tool_name: str) -> bool:
    tool_lower = tool_name.lower()
    if re.search(
        r"(?:login|log-?in|auth|signin|sign-in|verify|register|password|credential|account)",
        tool_lower,
        re.I,
    ):
        return True
    if re.search(r"\b(?:user(?:name)?|email|login|account|password)\b", blob, re.I):
        return True
    return any(p.search(blob) for p in USERNAME_ORACLE_PATTERNS)


def _probe_session_key(ctx: CallContext) -> str:
    tenant = ctx.tenant_id or os.environ.get("MASTYFF_AI_TENANT_ID") or "default"
    sub = "anon"
    if ctx.agent_identity:
        sub = ctx.agent_identity.sub or ctx.agent_identity.client_id or "anon"
    return f"{tenant}:{ctx.server_name}:{sub}"


def evaluate_timing_guard(ctx: CallContext) -> Optional[PolicyDecision]:
    if not is_timing_guard_enabled():
        return None

    blob = _build_timing_blob(ctx)
    if not blob.strip():
        return None

    session_key = _probe_session_key(ctx)
    matched, rule_ids = scan_timing_probe_patterns(blob)

    if matched:
        probe_count = _increment_counter(_probe_counters, session_key, PROBE_WINDOW_S)
        if probe_count > MAX_TIMING_PROBES:
            return PolicyDecision(
                "block",
                "timing-probe-rate-limit",
                f"Timing oracle probe rate exceeded ({probe_count}/{MAX_TIMING_PROBES} per minute)",
            )
        ids = ", ".join(rule_ids[:3])
        return PolicyDecision(
            "block",
            "timing-side-channel-guard",
            f"Timing side-channel probe detected ({ids})",
        )

    if _is_enumeration_probe_candidate(blob, ctx.tool_name):
        enum_key = f"{session_key}:{ctx.tool_name}:{enumeration_fingerprint(blob)}"
        enum_count = _increment_counter(_enum_counters, enum_key, ENUM_WINDOW_S)
        if enum_count > MAX_ENUM_PROBES:
            return PolicyDecision(
                "block",
                "timing-enumeration-guard",
                f"Timing enumeration oracle: {enum_count} similar probes for tool '{ctx.tool_name}'",
            )

    return None

File: src/policy/timing-guard.ts
/**
 * Enterprise timing side-channel mitigation — probe patterns, enumeration detection,
 * per-session rate limits, and constant-time pattern aggregation.
 */
import { createHash } from 'node:crypto';
import { LRUCache } from 'lru-cache';
import type { CallContext, PolicyDecision } from './policy-types.js';
import { walkStringLeaves } from './arg-leaf-walker.js';
import { deobfuscateRecursive } from '../utils/payload-normalizer.js';
import { stableFingerprint } from '../utils/constant-time.js';
import { envInt } from '../utils/eval-bounds.js';

/** Run every pattern; do not short-circuit on first match (reduces pattern-count timing leak). */
const TIMING_PROBE_PATTERNS: RegExp[] = [
  /\b(?:sleep|benchmark|pg_sleep|pg_sleep_for|pg_sleep_until|waitfor\s+delay|dbms_lock\.sleep|dbms_pipe\.receive_message|dbms_session\.sleep)\s*\(/i,
  /\bif\s*\(\s*(?:ascii|ord|substring|substr|mid|left|right)\s*\(/i,
  /\b(?:case\s+when|elt\s*\(|decode\s*\()\s+.*\b(?:sleep|benchmark|waitfor|pg_sleep)/i,
  /\b(?:timing|time[- ]?based)\s+(?:attack|oracle|injection|blind)/i,
  /\b(?:measure|detect|compare)\s+(?:response\s+)?time\s+(?:of|for|between)/i,
  /\b(?:valid|invalid)\s+username\b.*\b(?:timing|delay|sleep|benchmark)/i,
  /\busername\s+exists\b.*\b(?:time|delay|benchmark|sleep)/i,
  /\b(?:SLEEP|BENCHMARK)\s*\(\s*\d+/i,
  /\bWAITFOR\s+DELAY\s+'/i,
  /\b(?:select|union)\b.+\bwhere\b.+\b(?:sleep|benchmark|waitfor)/i,
  /\b'\s*or\s*'1'\s*=\s*'1\b.+\b(?:sleep|benchmark|waitfor|delay)/i,
  /\b(?:and|or)\s+\d+\s*=\s*\d+\s*--/i,
  /\bldap[_-]?search\b.+\b(?:delay|sleep|time)/i,
  /\$where\b.+\b(?:sleep|this\.constructor)/i,
  /\b(?:load_file|into\s+outfile)\b.+\b(?:sleep|benchmark)/i,
  /\bhex\s*\(\s*(?:substring|mid)\s*\(/i,
  /\b(?:response|elapsed)\s+time\s*(?:>|<|>=|<=)\s*\d+/i,
  /\buser[_-]?enumeration\b.*\b(?:timing|oracle)/i,
  /\b(?:binary|blind)\s+search\b.*\b(?:char|password|secret)/i,
];

const USERNAME_ORACLE_PATTERNS: RegExp[] = [
  /\bwhere\b\s+(?:user|username|email|login)\s*=\s*['"][^'"]{1,64}['"]/i,
  /\b(?:admin|root|administrator)\b.*\b(?:and|or)\b.*\b(?:sleep|benchmark|waitfor)/i,
  /\b(?:'|%27)\s*(?:or|and)\s*(?:'|%27)?\d+['"]?\s*=\s*['"]?\d+/i,
  /\b(?:exists|in\s*\(\s*select)\b.+\b(?:users|accounts|credentials)/i,
];

const PROBE_WINDOW_MS = 60_000;
const ENUM_WINDOW_MS = 120_000;

const MAX_TIMING_PROBES_PER_SESSION = envInt('MASTYFF_AI_MAX_TIMING_PROBES_PER_MIN', 8);
const MAX_ENUM_PROBES_PER_SESSION = envInt('MASTYFF_AI_MAX_ENUM_PROBES_PER_SESSION', 20);

const probeCounters = new LRUCache<string, { count: number; resetAt: number }>({
  max: 20_000,
  ttl: PROBE_WINDOW_MS,
});

const enumCounters = new LRUCache<string, { count: number; resetAt: number }>({
  max: 30_000,
  ttl: ENUM_WINDOW_MS,
});

export function isTimingGuardEnabled(): boolean {
  return process.env['MASTYFF_AI_TIMING_GUARD'] !== 'false';
}

function probeSessionKey(ctx: CallContext): string {
  const tenant = ctx.tenantId || process.env['MASTYFF_AI_TENANT_ID'] || 'default';
  const sub = ctx.agentIdentity?.sub || ctx.agentIdentity?.clientId || 'anon';
  return `${tenant}:${ctx.serverName}:${sub}`;
}

function buildTimingBlob(ctx: CallContext): string {
  return walkStringLeaves(ctx.arguments ?? {})
    .map((l) => deobfuscateRecursive(l.value))
    .join('\n');
}

/** Limit enumeration tracking to auth/username-oracle style probes (not benign path/query variance). */
function isEnumerationProbeCandidate(blob: string, toolName: string): boolean {
  const toolLower = toolName.toLowerCase();
  if (/(?:login|log-?in|auth|signin|sign-in|verify|register|password|credential|account)/i.test(toolLower)) {
    return true;
  }
  if (/\b(?:user(?:name)?|email|login|account|password)\b/i.test(blob)) {
    return true;
  }
  return USERNAME_ORACLE_PATTERNS.some((p) => p.test(blob));
}

/** Aggregate all pattern hits (no early exit). */
export function scanTimingProbePatterns(blob: string): { matched: boolean; ruleIds: string[] } {
  const ruleIds: string[] = [];
  if (!blob.trim()) return { matched: false, ruleIds };

  for (let i = 0; i < TIMING_PROBE_PATTERNS.length; i++) {
    if (TIMING_PROBE_PATTERNS[i].test(blob)) {
      ruleIds.push(`timing-probe-${i}`);
    }
  }
  for (let i = 0; i < USERNAME_ORACLE_PATTERNS.length; i++) {
    if (USERNAME_ORACLE_PATTERNS[i].test(blob)) {
      ruleIds.push(`username-oracle-${i}`);
    }
  }

  return { matched: ruleIds.length > 0, ruleIds };
}

/** Fingerprint for enumeration: collapse quoted strings and common usernames. */
export function enumerationFingerprint(blob: string): string {
  const normalized = blob
    .toLowerCase()
    .replace(/['"][^'"]{1,80}['"]/g, '<q>')
    .replace(/\b(?:admin|root|administrator|user\d*|test\d*|guest)\b/g, '<u>')
    .replace(/\d+/g, 'N');
  return stableFingerprint(normalized).slice(0, 16);
}

function incrementCounter(
  cache: LRUCache<string, { count: number; resetAt: number }>,
  key: string,
  windowMs: number,
): number {
  const now = Date.now();
  let counter = cache.get(key);
  if (!counter || now > counter.resetAt) {
    counter = { count: 1, resetAt: now + windowMs };
  } else {
    counter.count++;
  }
  cache.set(key, counter);
  return counter.count;
}

export function resetTimingProbeCounters(): void {
  probeCounters.clear();
  enumCounters.clear();
}

export function evaluateTimingGuard(ctx: CallContext): PolicyDecision | null {
  if (!isTimingGuardEnabled()) return null;

  const blob = buildTimingBlob(ctx);
  if (!blob.trim()) return null;

  const sessionKey = probeSessionKey(ctx);
  const scan = scanTimingProbePatterns(blob);

  if (scan.matched) {
    const probeCount = incrementCounter(probeCounters, sessionKey, PROBE_WINDOW_MS);
    if (probeCount > MAX_TIMING_PROBES_PER_SESSION) {
      return {
        action: 'block',
        rule: 'timing-probe-rate-limit',
        reason: `Timing oracle probe rate exceeded (${probeCount}/${MAX_TIMING_PROBES_PER_SESSION} per minute)`,
      };
    }
    return {
      action: 'block',
      rule: 'timing-side-channel-guard',
      reason: `Timing side-channel probe detected (${scan.ruleIds.slice(0, 3).join(', ')})`,
    };
  }

  if (isEnumerationProbeCandidate(blob, ctx.toolName)) {
    const enumKey = `${sessionKey}:${ctx.toolName}:${enumerationFingerprint(blob)}`;
    const enumCount = incrementCounter(enumCounters, enumKey, ENUM_WINDOW_MS);
    if (enumCount > MAX_ENUM_PROBES_PER_SESSION) {
      return {
        action: 'block',
        rule: 'timing-enumeration-guard',
        reason: `Timing enumeration oracle: ${enumCount} similar probes for tool '${ctx.toolName}'`,
      };
    }
  }

  return null;
}
